### PR TITLE
Fix bug in copying bytes for socket write

### LINKED ISSUE #45444: CopyBytesSocketChannel corrupts messages under load
In our nightly benchmarks we noticed response timeouts in several of the benchmarks since Friday's nightly run. Investigating the [range of commits](https://github.com/elastic/elasticsearch/compare/eb226773481a3b8c0354cb9827633e46c8527575...67a81ab95382c22b33a25c5d8ec4538024cdf793), we could narrow it down to https://github.com/elastic/elasticsearch/commit/e0f9d61becc45a470c97f3792872bdeb7fc9cae6. Starting with this commit we see corrupted responses like this (note the `"_seq_noc"` in the middle of the output):

```
{"index":{"_index":"weather-data-2016","_type":"_doc","_id":"mrbxhGwBLEbHlJDxHYLj","_version":1,"result":"created","_shards":{"total":1,"successful":1,"failed":0},"_seq_no":18764,"_primary_term":1,"status":201}},
{"index":{"_index":"weather-data-2016","_type":"_doc","_id":"m7bxhGwBLEbHlJDxHYLj","_version":1,"result":"created","_shards":{"total":1,"successful":1,"failed":0},"_seq_noc","_id":"CrbxhGwBLEbHlJDxHY3m","_version":1,"result":"created","_shards":{"total":1,"successful":1,"failed":0},"_seq_no":40023,"_primary_term":1,"status":201}}
```

**Steps to reproduce**:

This problem can be reproduced easily with Rally with the `noaa` track: 

```
esrally --track=noaa --challenge=append-no-conflicts --on-error=abort --revision=e0f9d61becc45a470c97f3792872bdeb7fc9cae6
```
Almost immediately after starting the benchmark, we get an error message along these lines (on the client):

```
JSONDecodeError("Expecting ':' delimiter: line 1 column 482447 (char 482446)
```

indicating corrupted JSON in the HTTP response.

**Additional notes**

* By reducing the bulk size from the default 5000 documents per bulk to e.g. 1000 documents with `--track-params="bulk_size:1000"` the benchmark finishes just fine so the problem does not seem to manifest with smaller requests / responses.

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we take the array of nio buffers from the netty channel
outbound buffer and copy their bytes to a direct buffer. In the process
we mutate the nio buffer positions. It seems like netty will continue to
reuse these buffers. This means than any data that is not flushed in a
call is lost. This commit fixes this by incrementing the positions after
the flush has completed. This is similar to the behavior that
SocketChannel would have provided and netty relied upon.

Fixes #45444.

### CODE REVIEW

**Comment by tbrooks8** [comment]
Probably also resolves the tests related to this issue: #45405.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java [hunk 1]
```diff
@@ -151,11 +149,20 @@ private void adjustMaxBytesPerGatheringWrite(int attempted, int written, int old
...
-            int initialLimit = buffer.limit();
-            buffer.limit(buffer.position() + nBytesToCopy);
-            destination.put(buffer);
-            buffer.limit(initialLimit);
+            destination.put(buffer.array(), buffer.arrayOffset() + buffer.position(), nBytesToCopy);
+        }
+    }
+
+    private static void setWrittenBytes(ByteBuffer[] source, int bytesWritten) {
+        int i = 0;
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java]: NIT: maybe keep the style we use in `copyBytes` for the loop to keep this a little shorter and easier to digest:

```
        for (int i = 0; bytesWritten > 0 && i < source.length; i++) {
            ByteBuffer buffer = source[i];
            int nBytes = Math.min(buffer.remaining(), bytesWritten);
            buffer.position(buffer.position() + nBytes);
            bytesWritten -= nBytes;
        }
```

**Review by original-brownbear — APPROVED** [review]
LGTM (we should add a test for writing large messages in a follow up IMO (obviously slightly tricky to reproduce on loopback ...), but lets get this in before we get a bunch more weird test failures :))

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java (hunk 1 above)
Comment (tbrooks8) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java]: I adopted this loop style without the `source.length` boundary condition. I don't think that boundary should be necessary as bytes written should not exceed the bytes in the buffers.

**Comment by tbrooks8** [comment]
> we should add a test for writing large messages in a follow up IMO (obviously slightly tricky to reproduce on loopback ...)

yeah. I had worked on an abstract simple ... test but could not get it to fail. My plan was to follow up with test for `CopyBytesSocketChannel` at minimum.

**Comment by original-brownbear** [comment]
@tbrooks8 I think you should be able to get a failing test by simply making it write to a socket that doesn't read for a while so that you get a partial write once that socket's buffers are full?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc
Jenkins run Elasticsearch-ci/default-distro

**Pull request merged**